### PR TITLE
[#390] event based comm example

### LINKED ISSUE #518: `WaitSet` goes into infinite loop when deadline was reached.
**Observed result or behaviour:**
When a deadline or an interval is attached to the `WaitSet`, the `WaitSet` waits for the provided time and then goes into an infinite loop retriggering the missed deadline/interval.

**Expected result or behaviour:**
When a deadline/interval with timeout X is attached, the `WaitSet` shall notify the user once when timeout X is reached and then reset the event. The next timeout missed notification shall be triggered after a timeout of X as again passed, so in total after 2*X.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Notes for Reviewer

 * Adds the Rust-version of a more complex event-driven communication example. The idea is that the publisher and subscriber emit explicitly a user-defined event whenever they do something like connect/disconnect/sent sample/receive sample/request history etc. Based on that example a C++ example will follow in a follow up PR. In the example the ports are part of a more generic `ExamplePublisher` or `ExampleSubscriber` that also contain an notifier/listener to wait or emit events that are related to them. It also provides a first preview how a high-level iceoryx2 API could look like, where the event notification and the message transport are no longer separated.
 * Fixes a Bug in the `WaitSet` where it went into an infinite loop after the first deadline/interval has passed. The only thing here missing was to store the last time `missed_deadlines` was called in the deadline queue.

## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked in the [References](#references) section
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

Relates to #390
Closes #518

### CODE REVIEW

File: iceoryx2/tests/waitset_tests.rs [hunk 1]
```diff
@@ -506,6 +506,26 @@ mod waitset {
...
+        let now = Instant::now();
+
+        std::thread::sleep(TIMEOUT + TIMEOUT / 10);
+
+        sut.wait_and_process_once(|_| CallbackProgression::Continue)
+            .unwrap();
+        sut.wait_and_process_once(|_| CallbackProgression::Continue)
+            .unwrap();
+
+        assert_that!(now.elapsed(), ge 2 * TIMEOUT);
```
Comment (elBoberido) [inline iceoryx2/tests/waitset_tests.rs:526]: This should be more robust in cast the first blocking wait is just way longer than expected and the second one does not wait at all.

```suggestion
        sut.wait_and_process_once(|_| CallbackProgression::Continue)
            .unwrap();
            
        std::thread::sleep(TIMEOUT + TIMEOUT / 10);
        
        sut.wait_and_process_once(|_| CallbackProgression::Continue)
            .unwrap();

        assert_that!(now.elapsed(), ge TIMEOUT);
```

File: examples/rust/_examples_common/pubsub_event.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+pub enum PubSubEvent {
+    PublisherConnected = 0,
+    PublisherDisconnected = 1,
+    SubscriberConnected = 2,
+    SubscriberDisconnected = 3,
+    SentSample = 4,
+    ReceivedSample = 5,
+    SentHistory = 6,
+    Undefined,
```
Comment (elBoberido) [inline examples/rust/_examples_common/pubsub_event.rs:23]: Food for thought
```suggestion
    Unknown,
```

File: examples/rust/event_based_communication/publisher.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    fn handle_event(&self) -> Result<(), Box<dyn std::error::Error>> {
+        while let Some(event) = self.listener.try_wait_one()? {
+            let event: PubSubEvent = event.into();
+            match event {
+                PubSubEvent::SubscriberConnected => {
+                    println!("new subscriber connected - delivering history");
+                    self.publisher.update_connections().unwrap();
+                    self.notifier
+                        .notify_with_custom_event_id(PubSubEvent::SentHistory.into())
+                        .unwrap();
```
Comment (elBoberido) [inline examples/rust/event_based_communication/publisher.rs:105]: Philosophical question. Should the `SentSample` event also be triggert?

File: examples/rust/event_based_communication/publisher.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (elBoberido) [inline examples/rust/event_based_communication/publisher.rs:1]:
```suggestion
// Copyright (c) 2024 Contributors to the Eclipse Foundation
```

File: examples/rust/event_based_communication/subscriber.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (elBoberido) [inline examples/rust/event_based_communication/subscriber.rs:1]:
```suggestion
// Copyright (c) 2024 Contributors to the Eclipse Foundation
```

**Review by elBoberido — COMMENTED** [review]
Looks good. Just some nitpicks.

What do you think about adding an FD based event to the example, e.g. for unix domain sockets.

File: examples/rust/event_based_communication/publisher.rs (hunk 3 above)
Comment (elfenpiff) [reply examples/rust/event_based_communication/publisher.rs:105]: I would not distinguish this. It is a simple example and the history may consist of N samples but then only 1 `SentSample` event was triggered which would be inconsistent.

File: iceoryx2/tests/waitset_tests.rs [hunk 6]
```diff
@@ -506,6 +506,27 @@ mod waitset {
...
+
+        sut.wait_and_process_once(|_| CallbackProgression::Continue)
+            .unwrap();
+
+        assert_that!(now.elapsed(), ge TIMEOUT);
+
+        sut.wait_and_process_once(|_| CallbackProgression::Continue)
+            .unwrap();
+
+        assert_that!(now.elapsed(), ge 2 * TIMEOUT);
```
Comment (elBoberido) [inline iceoryx2/tests/waitset_tests.rs:527]: Oh, I gave you bad advice. What I actually meant was this.

```suggestion
        std::thread::sleep(TIMEOUT + TIMEOUT / 10);
        
        sut.wait_and_process_once(|_| CallbackProgression::Continue)
            .unwrap();

        let now = Instant::now();

        sut.wait_and_process_once(|_| CallbackProgression::Continue)
            .unwrap();

        assert_that!(now.elapsed(), ge TIMEOUT / 2);
```

**Review by elBoberido — APPROVED** [review]

**Pull request merged**